File: gym_be/gym_be/Services/PaymentService.cs
using Microsoft.AspNetCore.Mvc;
using Stripe;
using Stripe.Checkout;
using Newtonsoft.Json;
using gym_be.Models;
using gym_be.Models.Entities;
using gym_be.Services.Interfaces;
using System.Globalization;
using System.Text.RegularExpressions;

namespace gym_be.Services
{
    public class PaymentService : IPaymentService
    {
        private readonly GymContext _context;

        public PaymentService(GymContext context)
        {
            _context = context;
        }

        public async Task<object> CreateCheckoutSessionAsync(CheckoutSessionRequest request)
        {
            StripeConfiguration.ApiKey = "sk_test_51Rom2g3PJbTWL2KKBiJZVnavBq2U1V1p7xrBWn1WI0Y5UM8xQ56LNXMmWMiY6AQMMk3vmXmdpinjVMQwYC20Logd00LGhiefLH";

            var origin = request.Origin ?? "http://localhost:3000";
            var successUrl = $"{origin}/user/cart?success=true";
            var cancelUrl = $"{origin}/user/cart?canceled=true";

            var options = new SessionCreateOptions
            {
                PaymentMethodTypes = new List<string> { "card" },
                LineItems = request.Items.Select(item => new SessionLineItemOptions
                {
                    PriceData = new SessionLineItemPriceDataOptions
                    {
                        UnitAmount = (long)(item.Price),
                        Currency = "vnd",
                        ProductData = new SessionLineItemPriceDataProductDataOptions
                        {
                            Name = item.Name,
                        },
                    },
                    Quantity = 1,
                }).ToList(),
                Mode = "payment",
                SuccessUrl = successUrl,
                CancelUrl = cancelUrl,
                Metadata = new Dictionary<string, string>
                {
                    { "customerid", request.CustomerId ?? "" },
                    { "courses", JsonConvert.SerializeObject(request.Items) }
                }
            };

            var service = new SessionService();
            Session session = service.Create(options);

            // Already camelCase
            return new { url = session.Url };
        }

        public async Task<object> SavePaymentAsync(SavePaymentRequest request)
        {
            var paymentId = Guid.NewGuid();
            var paidAt = DateTime.UtcNow;
            decimal total = request.Items.Sum(i => (decimal)(i.price));

            var payment = new Payment
            {
                PaymentId = paymentId,
                Amount = total,
                CustomerId = Guid.TryParse(request.CustomerId?.ToString(), out var cid) ? cid : (Guid?)null,
                Status = true,
                PaidAt = paidAt,
                Method = true
            };
            _context.Payments.Add(payment);

            foreach (var item in request.Items)
            {
                Guid courseGuid = Guid.TryParse(item.id?.ToString(), out var parsedId) ? parsedId : Guid.Empty;
                var detail = new PaymentDetail
                {
                    PaymentDetailId = Guid.NewGuid(),
                    PaymentId = paymentId,
                    CourseId = courseGuid,
                    Price = (decimal)item.price,
                    CustomerId = Guid.TryParse(request.CustomerId?.ToString(), out var cid2) ? cid2 : (Guid?)null
                };
                _context.PaymentDetails.Add(detail);
            }
            await _context.SaveChangesAsync();
            // Already camelCase
            return new { message = "Payment saved" };
        }

        public async Task<IEnumerable<object>> GetMyCoursesAsync(Guid customerId)
        {
            // Lấy các paymentId đã thanh toán thành công của customer
            var paidPaymentIds = _context.Payments
                .Where(p => p.CustomerId == customerId && p.Status == true)
                .Select(p => p.PaymentId)
                .ToList();

            // Lấy các khoá học đã mua, kèm tên dịch vụ (servicename) và tên PT (ptname)
            var courseList = (from pd in _context.PaymentDetails
                              join c in _context.WorkoutCourses on pd.CourseId equals c.CourseId
                              join s in _context.Services on c.ServiceId equals s.ServiceID into serviceJoin
                              from s in serviceJoin.DefaultIfEmpty()
                              join pt in _context.Customers on c.PersonalTrainerId equals pt.CustomerID into ptJoin
                              from pt in ptJoin.DefaultIfEmpty()
                              where pd.CustomerId == customerId && paidPaymentIds.Contains(pd.PaymentId)
                              select new {
                                  courseId = c.CourseId,
                                  courseName = c.CourseName,
                                  imageUrl = c.ImageUrl,
                                  personalTrainerId = c.PersonalTrainerId,
                                  durationWeek = c.DurationWeek,
                                  description = c.Description,
                                  price = c.Price,
                                  serviceId = c.ServiceId,
                                  serviceName = s != null ? s.ServiceName : null,
                                  ptName = pt != null ? pt.Name : null,
                                  schedules = c.Schedules
                              }).ToList();

            return courseList;
        }

        public async Task<IEnumerable<object>> GetMySchedulesAsync(Guid customerId)
        {
            // Lấy các paymentId đã thanh toán thành công của customer
            var paidPaymentIds = _context.Payments
                .Where(p => p.CustomerId == customerId && p.Status == true)
                .Select(p => p.PaymentId)
                .ToList();

            // Lấy các khoá học đã mua
            var courseIds = _context.PaymentDetails
                .Where(pd => pd.CustomerId == customerId && paidPaymentIds.Contains(pd.PaymentId))
                .Select(pd => pd.CourseId)
                .Distinct()
                .ToList();

            // Lấy tất cả WorkoutCourse liên quan
            var courses = _context.WorkoutCourses
                .Where(c => courseIds.Contains(c.CourseId))
                .AsEnumerable()
                .ToList();

            // Tạo ánh xạ scheduleId -> (PersonalTrainerId, CourseId)
            var scheduleCourseMap = new Dictionary<Guid, (Guid PersonalTrainerId, Guid CourseId)>();
            foreach (var course in courses)
            {
                foreach (var scheduleId in course.Schedules)
                {
                    scheduleCourseMap[scheduleId] = (course.PersonalTrainerId, course.CourseId);
                }
            }

            // Lấy tất cả scheduleId
            var scheduleIds = scheduleCourseMap.Keys.ToList();

            // Lấy thông tin chi tiết các schedule
            var schedules = _context.Schedules
                .Where(s => scheduleIds.Contains(s.ScheduleID))
                .ToList();

            // Lấy danh sách giáo viên
            var trainerIds = scheduleCourseMap.Values.Select(x => x.PersonalTrainerId).Distinct().ToList();
            var trainers = _context.Customers
                .Where(c => trainerIds.Contains(c.CustomerID))
                .ToDictionary(c => c.CustomerID, c => c.Name);

            // Trả về schedule kèm tên giáo viên và tên khoá học
            var result = courses.Select(c => new {
                teacherName = trainers.ContainsKey(c.PersonalTrainerId)
                    ? trainers[c.PersonalTrainerId]
                    : "Không rõ",
                courseId = c.CourseId,
                courseName = c.CourseName,
                courseStartDate = c.StartDate,
                courseEndDate = c.EndDate,
                duration = c.DurationWeek,
                schedules = schedules
                    .Where(s => c.Schedules.Contains(s.ScheduleID))
                    .Select(s => new {
                        scheduleId = s.ScheduleID,
                        dayOfWeek = s.DayOfWeek,
                        startTime = s.StartTime,
                        endTime = s.EndTime
                    })
                    .ToList()
            }).ToList();

            return result;
        }

        public async Task<IEnumerable<object>> GetPaymentHistoryAsync(Guid customerId)
        {
            var paymentHistory = (from p in _context.Payments
                                 join pd in _context.PaymentDetails on p.PaymentId equals pd.PaymentId
                                 join c in _context.WorkoutCourses on pd.CourseId equals c.CourseId
                                 join pt in _context.Customers on c.PersonalTrainerId equals pt.CustomerID into ptJoin
                                 from pt in ptJoin.DefaultIfEmpty()
                                 where p.CustomerId == customerId
                                 orderby p.PaidAt descending
                                 select new
                                 {
                                     paymentId = p.PaymentId.ToString(),
                                     courseId = c.CourseId.ToString(),
                                     courseName = c.CourseName,
                                     instructor = pt != null ? pt.Name : "Không rõ",
                                     amount = pd.Price,
                                     originalAmount = c.Price,
                                     status = p.Status == true ? "completed" : "failed",
                                     date = p.PaidAt.HasValue ? p.PaidAt.Value.ToString("yyyy-MM-dd") : DateTime.Now.ToString("yyyy-MM-dd"),
                                     paymentMethod = p.Method == true ? "Thẻ tín dụng" : "Chuyển khoản",
                                     transactionId = p.PaymentId.ToString(), // Sử dụng paymentId thay cho mã giao dịch
                                    //  discount = c.Price - pd.Price,
                                     paidAt = p.PaidAt
                                 }).ToList();
            var result = paymentHistory.Cast<object>().ToList();
            return result;
        }
    }
} 

File: gym_fe/src/service/paymentService.ts
// src/service/paymentService.ts

interface CheckoutItem {
  Name: string;
  Price: number;
}

interface CheckoutSessionRequest {
  Items: CheckoutItem[];
  Origin: string;
  CustomerId: string;
}

interface SavePaymentItem {
  id: string;
  coursename: string;
  price: number;
}

interface SavePaymentRequest {
  Items: SavePaymentItem[];
  CustomerId: string;
}

interface PaymentResponse {
  url?: string;
  message?: string;
}

interface Course {
  courseid: string;
  coursename: string;
  imageurl: string;
  personaltrainerid: string;
  durationweek: number;
  description: string;
  price: number;
  serviceid: string;
  schedules: string[];
}
export interface Schedule {
  scheduleId: string;
  dayOfWeek: string;
  maxParticipants: number;
  startTime: string;
  endTime: string;
  courseId: string;
  courseName: string;
  teacherName: string;
}

export type ScheduleDetail = {
  scheduleId: string;
  dayOfWeek: 'Chủ nhật' | 'Thứ 2' | 'Thứ 3' | 'Thứ 4' | 'Thứ 5' | 'Thứ 6' | 'Thứ 7';
  startTime: string; // Format: 'HH:mm'
  endTime: string;   // Format: 'HH:mm'
};

export type ScheduleResponse = {
  courseId: string;
  teacherName: string;
  courseName: string;
  courseStartDate: string; // ISO format date string
  courseEndDate: string;   // ISO format date string
  duration: number;        // in weeks
  schedules: Schedule[];
};
interface PaymentHistory {
  paymentId: string;
  courseId: string;
  courseName: string;
  instructor: string;
  amount: number;
  originalAmount: number;
  status: string;
  date: string;
  paymentMethod: string;
  transactionId: string;
  discount: number;
  paidAt: string;
}

class PaymentService {
  private baseUrl = 'http://localhost:5231/api/payment';

  async createCheckoutSession(request: CheckoutSessionRequest): Promise<PaymentResponse> {
    const response = await fetch(`${this.baseUrl}/create-checkout-session`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to create checkout session');
    }

    return response.json();
  }

  async savePayment(request: SavePaymentRequest): Promise<PaymentResponse> {
    const response = await fetch(`${this.baseUrl}/save`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to save payment');
    }

    return response.json();
  }

  async getMyCourses(customerId: string): Promise<Course[]> {
    const response = await fetch(`${this.baseUrl}/my-courses/${customerId}`, {
      method: 'GET',
      headers: { 'Content-Type': 'application/json' },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to get my courses');
    }

    return response.json();
  }

  async getMySchedules(customerId: string): Promise<ScheduleResponse[]> {
    const response = await fetch(`${this.baseUrl}/my-schedules/${customerId}`, {
      method: 'GET',
      headers: { 'Content-Type': 'application/json' },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to get my schedules');
    }

    return response.json();
  }

  async getPaymentHistory(customerId: string): Promise<PaymentHistory[]> {
    const response = await fetch(`${this.baseUrl}/history/${customerId}`, {
      method: 'GET',
      headers: { 'Content-Type': 'application/json' },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to get payment history');
    }

    return response.json();
  }

  async getUserStats(customerId: string): Promise<{ totalPackages: number; totalSpent: number }> {
    const response = await fetch('http://localhost:5231/api/dashboard/user-stats/' + customerId, {
      method: 'GET',
      headers: { 'Content-Type': 'application/json' },
    });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to get user stats');
    }
    return response.json();
  }
}


export default new PaymentService();
export type { PaymentHistory }; 